File: js-ts-automation/playwrite/Pages/CheckoutPage.ts
import { Page, expect } from '@playwright/test';
import { URLS } from '../data/testData';

export interface CustomerInfo {
  firstName: string;
  lastName: string;
  postalCode: string;
}

/**
 * CheckoutPage — covers Step 1 (info), Step 2 (overview), and confirmation.
 */
export class CheckoutPage {
  private readonly page: Page;

  readonly firstNameInput = () => this.page.getByTestId('firstName');
  readonly lastNameInput = () => this.page.getByTestId('lastName');
  readonly postalCodeInput = () => this.page.getByTestId('postalCode');
  readonly continueButton = () => this.page.getByTestId('continue');
  readonly cancelButton = () => this.page.getByTestId('cancel');
  readonly step1ErrorMessage = () => this.page.getByTestId('error');

  readonly subtotalLabel = () => this.page.locator('.summary_subtotal_label');
  readonly taxLabel = () => this.page.locator('.summary_tax_label');
  readonly totalLabel = () => this.page.locator('.summary_total_label');
  readonly finishButton = () => this.page.getByTestId('finish');

  readonly completeHeader = () => this.page.getByTestId('complete-header');
  readonly completeText = () => this.page.getByTestId('complete-text');
  readonly backHomeButton = () => this.page.getByTestId('back-to-products');

  constructor(page: Page) {
    this.page = page;
  }

  // ─── Step 1 actions ─────────────────────────────────────────────────────────

  async fillCustomerInfo(info: CustomerInfo): Promise<void> {
    await this.firstNameInput().fill(info.firstName);
    await this.lastNameInput().fill(info.lastName);
    await this.postalCodeInput().fill(info.postalCode);
  }

  async continueToOverview(): Promise<void> {
    await this.continueButton().click();
    await expect(this.page).toHaveURL(URLS.checkoutStep2);
  }

  async fillAndContinue(info: CustomerInfo): Promise<void> {
    await this.fillCustomerInfo(info);
    await this.continueToOverview();
  }

  /** Submit empty form to trigger validation; returns so callers can chain */
  async submitEmptyForm(): Promise<void> {
    await this.continueButton().click();
  }

  // ─── Step 2 actions ─────────────────────────────────────────────────────────

  async finishOrder(): Promise<void> {
    await this.finishButton().click();
    await expect(this.page).toHaveURL(URLS.checkoutComplete);
  }

  /** Returns { subtotal, tax, total } as numbers for math verification */
  async getPriceSummary(): Promise<{ subtotal: number; tax: number; total: number }> {
    const subtotalText = (await this.subtotalLabel().textContent()) ?? '';
    const taxText = (await this.taxLabel().textContent()) ?? '';
    const totalText = (await this.totalLabel().textContent()) ?? '';

    return {
      subtotal: parseFloat(subtotalText.replace('Item total: $', '').trim()),
      tax: parseFloat(taxText.replace('Tax: $', '').trim()),
      total: parseFloat(totalText.replace('Total: $', '').trim()),
    };
  }

  // ─── Assertions ──────────────────────────────────────────────────────────────

  async assertStep1Error(errorText: string): Promise<void> {
    await expect(this.step1ErrorMessage()).toBeVisible();
    await expect(this.step1ErrorMessage()).toContainText(errorText);
  }

  /** Core financial assertion: subtotal + tax must equal total */
  async assertTotalCalculationIsCorrect(): Promise<void> {
    const { subtotal, tax, total } = await this.getPriceSummary();
    expect(
      subtotal + tax,
      `Expected subtotal (${subtotal}) + tax (${tax}) = ${subtotal + tax}, got ${total}`
    ).toBeCloseTo(total, 2);
  }

  async assertConfirmationPage(): Promise<void> {
    await expect(this.page).toHaveURL(URLS.checkoutComplete);
    await expect(this.completeHeader()).toBeVisible();
    await expect(this.completeHeader()).toHaveText('Thank you for your order!');
  }

  async assertOrderConfirmationDetails(): Promise<void> {
    await this.assertConfirmationPage();
    await expect(this.completeText()).toContainText('Your order has been dispatched');
    await expect(this.backHomeButton()).toBeVisible();
  }
}

File: reqnroll-automation/playwright/Pages/CheckoutPage.cs
using System.Globalization;
using Microsoft.Playwright;
using NUnit.Framework;
using ReqnrollPlaywright.Fixtures;

namespace ReqnrollPlaywright.Pages;

/// <summary>
/// CheckoutPage — covers Step 1 (info), Step 2 (overview), and confirmation.
/// </summary>
public class CheckoutPage
{
    private readonly IPage _page;

    public CheckoutPage(IPage page)
    {
        _page = page;
    }

    public ILocator FirstNameInput => _page.Locator("[data-test='firstName']");
    public ILocator LastNameInput => _page.Locator("[data-test='lastName']");
    public ILocator PostalCodeInput => _page.Locator("[data-test='postalCode']");
    public ILocator ContinueButton => _page.Locator("[data-test='continue']");
    public ILocator CancelButton => _page.Locator("[data-test='cancel']");
    public ILocator ErrorMessage => _page.Locator("[data-test='error']");

    public ILocator SubtotalLabel => _page.Locator(".summary_subtotal_label");
    public ILocator TaxLabel => _page.Locator(".summary_tax_label");
    public ILocator TotalLabel => _page.Locator(".summary_total_label");
    public ILocator FinishButton => _page.Locator("[data-test='finish']");

    public ILocator CompleteHeader => _page.Locator("[data-test='complete-header']");
    public ILocator CompleteText => _page.Locator("[data-test='complete-text']");
    public ILocator BackHomeButton => _page.Locator("[data-test='back-to-products']");

    // --- Step 1 actions -----------------------------------------------------

    public async Task FillCustomerInfoAsync(string firstName, string lastName, string postalCode)
    {
        await FirstNameInput.FillAsync(firstName);
        await LastNameInput.FillAsync(lastName);
        await PostalCodeInput.FillAsync(postalCode);
    }

    public async Task ContinueToOverviewAsync()
    {
        await ContinueButton.ClickAsync();
        Assert.That(_page.Url, Is.EqualTo(Urls.CheckoutStep2));
    }

    public async Task FillAndContinueAsync(string firstName, string lastName, string postalCode)
    {
        await FillCustomerInfoAsync(firstName, lastName, postalCode);
        await ContinueToOverviewAsync();
    }

    /// <summary>Submit empty form to trigger validation; does not assert navigation.</summary>
    public async Task SubmitEmptyFormAsync()
    {
        await ContinueButton.ClickAsync();
    }

    // --- Step 2 actions -----------------------------------------------------

    public async Task FinishOrderAsync()
    {
        await FinishButton.ClickAsync();
        Assert.That(_page.Url, Is.EqualTo(Urls.CheckoutComplete));
    }

    /// <summary>Returns (Subtotal, Tax, Total) as numbers for math verification.</summary>
    public async Task<(double Subtotal, double Tax, double Total)> GetPriceSummaryAsync()
    {
        var subtotalText = await SubtotalLabel.TextContentAsync() ?? string.Empty;
        var taxText = await TaxLabel.TextContentAsync() ?? string.Empty;
        var totalText = await TotalLabel.TextContentAsync() ?? string.Empty;

        var subtotal = ParseAmount(subtotalText, "Item total: $");
        var tax = ParseAmount(taxText, "Tax: $");
        var total = ParseAmount(totalText, "Total: $");

        return (subtotal, tax, total);
    }

    // --- Assertions ---------------------------------------------------------

    public async Task AssertStep1ErrorAsync(string expectedText)
    {
        Assert.That(await ErrorMessage.IsVisibleAsync(), Is.True);
        Assert.That(await ErrorMessage.TextContentAsync(), Does.Contain(expectedText));
    }

    /// <summary>Core financial assertion: subtotal + tax must equal total (within 0.01).</summary>
    public async Task AssertTotalCalculationIsCorrectAsync()
    {
        var (subtotal, tax, total) = await GetPriceSummaryAsync();
        Assert.That(subtotal + tax, Is.EqualTo(total).Within(0.01),
            $"Expected subtotal ({subtotal}) + tax ({tax}) = {subtotal + tax}, got {total}");
    }

    public async Task AssertConfirmationPageAsync()
    {
        Assert.That(_page.Url, Is.EqualTo(Urls.CheckoutComplete));
        Assert.That(await CompleteHeader.IsVisibleAsync(), Is.True);
        Assert.That(await CompleteHeader.TextContentAsync(), Is.EqualTo("Thank you for your order!"));
    }

    private static double ParseAmount(string text, string prefix)
    {
        var cleaned = text.Replace(prefix, string.Empty).Trim();
        return double.Parse(cleaned, CultureInfo.InvariantCulture);
    }
}
